File: mobile-native/android/ScDecoder.kt
package com.scdataminifier.decoder

import android.graphics.Bitmap

/**
 * Decode-only image SDK (Android). Turns a stored IMAGE value — the 1-byte descriptor plus the
 * version-1 headerless data carried in an SCDataMinifier TLV — into an Android [Bitmap], gated
 * by a license. The underlying image format is never named here or in the native library; the
 * app links whichever flavour (A or B) matches the images it must display.
 *
 * Packaging: put the flavour's `libscdec.so` under `src/main/jniLibs/<abi>/`. Both flavours use
 * the same class/symbol names, so an app links exactly one.
 *
 * ```
 * ScDecoder.license(licBytes, BuildConfig.APPLICATION_ID)   // once, at startup
 * val bmp = ScDecoder.decode(imageValueFromTlv)             // per image
 * ```
 */
object ScDecoder {

    init {
        System.loadLibrary("scdec")
    }

    /** Applies a signed license bound to [packageName]; returns 0 on success, negative otherwise. */
    fun license(license: ByteArray, packageName: String): Int = nLicense(license, packageName)

    /** True once a valid license has been accepted. */
    val isLicensed: Boolean get() = nLicensed()

    /** Pixel dimensions of an IMAGE value without decoding, or null on failure. */
    fun size(imageValue: ByteArray): Pair<Int, Int>? =
        nInfo(imageValue)?.let { it[0] to it[1] }

    /**
     * Decodes an IMAGE value to an ARGB_8888 [Bitmap] (opaque), or null on failure
     * (e.g. not licensed, or malformed value).
     */
    fun decode(imageValue: ByteArray): Bitmap? {
        val packed = nOpen(imageValue) ?: return null
        val w = ((packed[0].toInt() and 0xFF) shl 8) or (packed[1].toInt() and 0xFF)
        val h = ((packed[2].toInt() and 0xFF) shl 8) or (packed[3].toInt() and 0xFF)
        val pixels = IntArray(w * h)
        var p = 4
        for (i in 0 until w * h) {
            val r = packed[p].toInt() and 0xFF
            val g = packed[p + 1].toInt() and 0xFF
            val b = packed[p + 2].toInt() and 0xFF
            pixels[i] = (0xFF shl 24) or (r shl 16) or (g shl 8) or b
            p += 3
        }
        return Bitmap.createBitmap(pixels, w, h, Bitmap.Config.ARGB_8888)
    }

    private external fun nLicense(license: ByteArray, packageName: String): Int
    private external fun nLicensed(): Boolean
    private external fun nInfo(value: ByteArray): IntArray?
    private external fun nOpen(value: ByteArray): ByteArray?
}

File: mobile-native/ios/ScDecoder.swift
import Foundation
import CoreGraphics

/// Decode-only image SDK (iOS). Turns a stored IMAGE value — the 1-byte descriptor plus the
/// version-1 headerless data carried in an SCDataMinifier TLV — into a `CGImage`, gated by a
/// license. The underlying image format is never named here or in the embedded `scdec.xcframework`;
/// the app embeds whichever flavour (A or B) matches the images it must display.
///
/// The framework exposes the C API declared in `scdec.h` (module `scdec`). Import it and use:
/// ```
/// ScDecoder.license(licData, packageName: Bundle.main.bundleIdentifier!)  // once
/// let cg = ScDecoder.decode(imageValue)                                   // per image
/// ```
public enum ScDecoder {

    @discardableResult
    public static func license(_ license: Data, packageName: String) -> Int {
        license.withUnsafeBytes { raw in
            Int(scdec_license(raw.bindMemory(to: UInt8.self).baseAddress, Int32(license.count), packageName))
        }
    }

    public static var isLicensed: Bool { scdec_licensed() != 0 }

    /// Pixel dimensions of an IMAGE value without decoding, or nil on failure.
    public static func size(_ value: Data) -> (width: Int, height: Int)? {
        var w: Int32 = 0, h: Int32 = 0
        let ok = value.withUnsafeBytes { raw in
            scdec_info(raw.bindMemory(to: UInt8.self).baseAddress, Int32(value.count), &w, &h)
        }
        return ok == 0 ? (Int(w), Int(h)) : nil
    }

    /// Decodes an IMAGE value to an opaque RGB `CGImage`, or nil on failure.
    public static func decode(_ value: Data) -> CGImage? {
        var w: Int32 = 0, h: Int32 = 0
        guard let rgb = value.withUnsafeBytes({ raw in
            scdec_open(raw.bindMemory(to: UInt8.self).baseAddress, Int32(value.count), &w, &h)
        }) else { return nil }
        let width = Int(w), height = Int(h), count = width * height * 3
        defer { scdec_free(rgb) }

        let data = Data(bytes: rgb, count: count)
        guard let provider = CGDataProvider(data: data as CFData) else { return nil }
        return CGImage(
            width: width, height: height,
            bitsPerComponent: 8, bitsPerPixel: 24, bytesPerRow: width * 3,
            space: CGColorSpaceCreateDeviceRGB(),
            bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.none.rawValue),
            provider: provider, decode: nil, shouldInterpolate: true, intent: .defaultIntent)
    }
}
